node/core/candidate-validation/src/lib.rs

```diff
@@ -196,73 +236,58 @@ where
 }
 
 #[derive(Debug)]
-enum AssumptionCheckOutcome {
-	Matches(PersistedValidationData, ValidationCode),
+enum AssumptionCheckOutcome<T> {
+	Matches(T, OccupiedCoreAssumption),
 	DoesNotMatch,
 	BadRequest,
 }
 
-async fn check_assumption_validation_data<Context>(
+async fn check_assumption_validation_data<Context, T>(
```

Comment:
TBH, I am not really a fan of this approach - i.e. using the trait and trying to reuse this code. Sure, it perhaps minimizes the diff but it does so at the cost of legibility. If possible I would prefer straight line code.

Luckily, I think this should be possible to avoid.

Specifically, look at the `CandidateDescriptor`. It contains the hash of validation code against which the candidate was produced. This enables us to avoid querying the validation code before and allows us to query the validation code by the hash when it is needed. No need to try to match it against the assumption or anything alike.